File: backend/app/models/laboratorio.py
"""Laboratorio models: mediciones, clasificacion_cluster, umbrales_calidad,
registros_fenologicos, ejecucion_labores.
"""

from datetime import datetime, date
from decimal import Decimal
from typing import Optional
from sqlmodel import SQLModel, Field, Column
import sqlalchemy as sa


class MedicionLaboratorio(SQLModel, table=True):
    __tablename__ = "mediciones_laboratorio"
    id_medicion: Optional[int] = Field(default=None, primary_key=True)
    id_posicion: Optional[int] = Field(default=None, foreign_key="posiciones_testblock.id_posicion")
    id_planta: Optional[int] = Field(default=None, foreign_key="plantas.id_planta")
    temporada: Optional[str] = Field(default=None, sa_column=Column(sa.String(50)))
    fecha_medicion: date
    fecha_cosecha: Optional[date] = Field(default=None)
    brix: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 2)))
    acidez: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 3)))
    firmeza: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 1)))
    calibre: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 2)))
    peso: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(6, 2)))
    color_pct: Optional[int] = Field(default=None)
    cracking_pct: Optional[int] = Field(default=None)
    observaciones: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(None)))
    usuario_registro: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    fecha_creacion: Optional[datetime] = Field(default_factory=datetime.utcnow)

    # Firmeza detallada (5 puntos de medicion)
    firmeza_punta: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(8, 2)))
    firmeza_quilla: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(8, 2)))
    firmeza_hombro: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(8, 2)))
    firmeza_mejilla_1: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(8, 2)))
    firmeza_mejilla_2: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(8, 2)))

    # Muestra y postcosecha
    n_muestra: Optional[int] = Field(default=None)
    periodo_almacenaje: Optional[int] = Field(default=None)
    perimetro: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(8, 2)))
    pardeamiento: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 2)))
    traslucidez: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 2)))
    gelificacion: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 2)))
    harinosidad: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 2)))
    color_pulpa: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))

    # Agronomia y contexto de muestra
    raleo_frutos: Optional[int] = Field(default=None)
    rendimiento: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(10, 2)))
    repeticion: Optional[int] = Field(default=None)

    # Color de cubrimiento (% de frutos en cada rango)
    color_0_30: Optional[int] = Field(default=None)
    color_30_50: Optional[int] = Field(default=None)
    color_50_75: Optional[int] = Field(default=None)
    color_75_100: Optional[int] = Field(default=None)
    color_total: Optional[int] = Field(default=None)

    # Distribucion de color (conteo o % de frutos por categoria)
    color_verde: Optional[int] = Field(default=None)
    color_crema: Optional[int] = Field(default=None)
    color_amarillo: Optional[int] = Field(default=None)
    color_full: Optional[int] = Field(default=None)
    color_dist_total: Optional[int] = Field(default=None)

    # Total de frutos evaluados por metrica de postcosecha
    total_frutos_pardeamiento: Optional[int] = Field(default=None)
    total_frutos_traslucidez: Optional[int] = Field(default=None)
    total_frutos_gelificacion: Optional[int] = Field(default=None)
    total_frutos_harinosidad: Optional[int] = Field(default=None)

    # FKs directas para contexto de laboratorio
    id_campo: Optional[int] = Field(default=None, foreign_key="campos.id_campo")
    id_variedad: Optional[int] = Field(default=None, foreign_key="variedades.id_variedad")
    id_especie: Optional[int] = Field(default=None, foreign_key="especies.id_especie")
    id_portainjerto: Optional[int] = Field(default=None, foreign_key="portainjertos.id_portainjerto")


class ClasificacionCluster(SQLModel, table=True):
    __tablename__ = "clasificacion_cluster"
    id_clasificacion: Optional[int] = Field(default=None, primary_key=True)
    id_medicion: Optional[int] = Field(default=None, foreign_key="mediciones_laboratorio.id_medicion")
    cluster: Optional[int] = Field(default=None)
    banda_brix: Optional[int] = Field(default=None)
    banda_firmeza: Optional[int] = Field(default=None)
    banda_acidez: Optional[int] = Field(default=None)
    banda_calibre: Optional[int] = Field(default=None)
    score_total: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 2)))
    metodo: Optional[str] = Field(default=None, sa_column=Column(sa.String(20)))
    fecha_calculo: Optional[datetime] = Field(default_factory=datetime.utcnow)


class UmbralCalidad(SQLModel, table=True):
    __tablename__ = "umbrales_calidad"
    id_umbral: Optional[int] = Field(default=None, primary_key=True)
    id_especie: int = Field(foreign_key="especies.id_especie")
    metrica: str = Field(sa_column=Column(sa.String(30), nullable=False))
    banda: int
    valor_min: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(10, 3)))
    valor_max: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(10, 3)))
    peso_ponderacion: Optional[Decimal] = Field(default=None, sa_column=Column(sa.DECIMAL(5, 2)))
    activo: Optional[bool] = Field(default=True)
    fecha_modificacion: Optional[datetime] = Field(default=None)


class RegistroFenologico(SQLModel, table=True):
    __tablename__ = "registros_fenologicos"
    id_registro: Optional[int] = Field(default=None, primary_key=True)
    id_posicion: Optional[int] = Field(default=None, foreign_key="posiciones_testblock.id_posicion")
    id_planta: Optional[int] = Field(default=None, foreign_key="plantas.id_planta")
    id_estado_fenol: Optional[int] = Field(default=None, foreign_key="estados_fenologicos.id_estado")
    temporada: Optional[str] = Field(default=None, sa_column=Column(sa.String(20)))
    fecha_registro: date
    porcentaje: Optional[int] = Field(default=None)
    observaciones: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(None)))
    foto_url: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(500)))
    usuario_registro: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    fecha_creacion: Optional[datetime] = Field(default_factory=datetime.utcnow)


class DetalleLabor(SQLModel, table=True):
    """Sub-items / instructions for each tipo_labor, optionally filtered by species."""
    __tablename__ = "detalles_labor"
    id_detalle: Optional[int] = Field(default=None, primary_key=True)
    id_labor: int = Field(foreign_key="tipos_labor.id_labor")
    descripcion: str = Field(sa_column=Column(sa.NVARCHAR(500), nullable=False))
    aplica_especie: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(100)))
    orden: Optional[int] = Field(default=0)
    es_checklist: Optional[bool] = Field(default=True)
    activo: Optional[bool] = Field(default=True)
    fecha_creacion: Optional[datetime] = Field(default_factory=datetime.utcnow)
    usuario_creacion: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(100)))
    fecha_modificacion: Optional[datetime] = Field(default=None)
    usuario_modificacion: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(100)))


class EjecucionLabor(SQLModel, table=True):
    __tablename__ = "ejecucion_labores"
    __table_args__ = {"extend_existing": True}

    id_ejecucion: Optional[int] = Field(default=None, primary_key=True)
    id_posicion: Optional[int] = Field(default=None, foreign_key="posiciones_testblock.id_posicion")
    id_planta: Optional[int] = Field(default=None, foreign_key="plantas.id_planta")
    id_labor: Optional[int] = Field(default=None, foreign_key="tipos_labor.id_labor")
    temporada: Optional[str] = Field(default=None, sa_column=Column(sa.String(20)))
    fecha_programada: Optional[date] = Field(default=None)
    fecha_ejecucion: Optional[date] = Field(default=None)
    estado: Optional[str] = Field(default="ejecutada", sa_column=Column(sa.String(20)))
    ejecutor: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(100)))
    duracion_min: Optional[int] = Field(default=None)
    observaciones: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(None)))
    id_orden_trabajo: Optional[int] = Field(default=None, foreign_key="ordenes_trabajo.id")
    id_lote: Optional[int] = Field(default=None, foreign_key="inventario_vivero.id_inventario")
    usuario_registro: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    fecha_creacion: Optional[datetime] = Field(default_factory=datetime.utcnow)

File: frontend/src/types/laboratorio.ts
export interface MedicionLaboratorio {
  id_medicion: number;
  id_posicion?: number | null;
  id_planta?: number | null;
  temporada?: string | null;
  fecha_medicion: string;
  fecha_cosecha?: string | null;
  brix?: number | null;
  acidez?: number | null;
  firmeza?: number | null;
  calibre?: number | null;
  peso?: number | null;
  color_pct?: number | null;
  cracking_pct?: number | null;
  observaciones?: string | null;
  usuario_registro?: string | null;
  fecha_creacion?: string;
  // Firmeza detallada (5 puntos)
  firmeza_punta?: number | null;
  firmeza_quilla?: number | null;
  firmeza_hombro?: number | null;
  firmeza_mejilla_1?: number | null;
  firmeza_mejilla_2?: number | null;
  // Muestra y postcosecha
  n_muestra?: number | null;
  periodo_almacenaje?: number | null;
  perimetro?: number | null;
  pardeamiento?: number | null;
  traslucidez?: number | null;
  gelificacion?: number | null;
  harinosidad?: number | null;
  color_pulpa?: string | null;
  // Agronomia y contexto de muestra
  raleo_frutos?: number | null;
  rendimiento?: number | null;
  repeticion?: number | null;
  // Color de cubrimiento (% frutos en cada rango)
  color_0_30?: number | null;
  color_30_50?: number | null;
  color_50_75?: number | null;
  color_75_100?: number | null;
  color_total?: number | null;
  // Distribución de color
  color_verde?: number | null;
  color_crema?: number | null;
  color_amarillo?: number | null;
  color_full?: number | null;
  color_dist_total?: number | null;
  // Total frutos evaluados por metrica postcosecha
  total_frutos_pardeamiento?: number | null;
  total_frutos_traslucidez?: number | null;
  total_frutos_gelificacion?: number | null;
  total_frutos_harinosidad?: number | null;
  // FKs directas
  id_campo?: number | null;
  id_variedad?: number | null;
  id_especie?: number | null;
  id_portainjerto?: number | null;
}

export interface PaginatedMediciones {
  data: MedicionLaboratorio[];
  total: number;
  skip: number;
  limit: number;
}

export interface MedicionCreateResponse {
  medicion: MedicionLaboratorio;
  clasificacion: ClasificacionResult | null;
}

export interface ClasificacionResult {
  cluster: number;
  cluster_label: string;
  banda_brix: number;
  banda_firmeza: number;
  banda_acidez: number;
  banda_firmeza_punto: number;
  score_total: number | null;
}

export interface ClasificacionCluster {
  id_clasificacion: number;
  id_medicion: number;
  cluster?: number | null;
  banda_brix?: number | null;
  banda_firmeza?: number | null;
  banda_acidez?: number | null;
  banda_calibre?: number | null;
  score_total?: number | null;
  metodo?: string | null;
  fecha_calculo?: string;
}

export interface BatchRowResult {
  index: number;
  success: boolean;
  medicion?: {
    id_medicion: number;
    id_planta?: number | null;
    id_posicion?: number | null;
    temporada?: string | null;
    fecha_medicion?: string | null;
    brix?: number | null;
    acidez?: number | null;
    firmeza?: number | null;
    peso?: number | null;
  } | null;
  clasificacion?: ClasificacionResult | null;
  error?: string | null;
}

export interface BatchResponse {
  total_enviadas: number;
  total_creadas: number;
  total_errores: number;
  resultados: BatchRowResult[];
}

export interface LabKpis {
  total: number;
  brix_promedio: number | null;
  firmeza_promedio: number | null;
  acidez_promedio: number | null;
  calibre_promedio: number | null;
  brix_min: number | null;
  brix_max: number | null;
}

export interface Planta {
  id_planta: number;
  codigo?: string | null;
  id_posicion?: number | null;
  id_variedad?: number | null;
  id_portainjerto?: number | null;
  id_especie?: number | null;
  id_pmg?: number | null;
  id_lote_origen?: number | null;
  condicion: string;
  activa: boolean;
  ano_plantacion?: number | null;
}

export interface VariedadResumen {
  id_variedad: number;
  id_portainjerto: number | null;
  id_campo: number | null;
  id_pmg: number | null;
  variedad: string;
  especie: string;
  portainjerto: string;
  pmg: string;
  campo: string;
  total_mediciones: number;
  n_temporadas: number;
  brix_avg: number | null;
  brix_min: number | null;
  brix_max: number | null;
  firmeza_avg: number | null;
  acidez_avg: number | null;
  peso_avg: number | null;
  calibre_avg: number | null;
  cluster_dist: { c1: number; c2: number; c3: number; c4: number };
  cluster_predominante: number | null;
}

export interface EvolucionTemporada {
  temporada: string;
  total: number;
  brix_avg: number | null;
  firmeza_avg: number | null;
  acidez_avg: number | null;
  peso_avg: number | null;
  cluster_dist: { c1: number; c2: number; c3: number; c4: number };
}

export interface MedicionIndividual {
  id: number;
  fecha: string | null;
  temporada: string;
  brix: number | null;
  firmeza: number | null;
  acidez: number | null;
  peso: number | null;
  calibre: number | null;
  n_muestra: number | null;
  id_campo: number | null;
  campo: string;
  cluster: number | null;
}

export interface EvolucionResponse {
  contexto: {
    variedad: string;
    especie: string;
    portainjerto: string;
    pmg: string;
    campo: string | null;
  };
  por_temporada: EvolucionTemporada[];
  por_fecha: MedicionIndividual[];
}

export interface EjecucionLabor {
  id_ejecucion: number;
  id_posicion?: number | null;
  id_planta?: number | null;
  id_labor: number;
  temporada?: string | null;
  fecha_programada?: string | null;
  fecha_ejecucion?: string | null;
  estado: string;
  ejecutor?: string | null;
  duracion_min?: number | null;
  observaciones?: string | null;
  usuario_registro?: string | null;
  fecha_creacion?: string;
}
